api/v1/tlsroute.go

```diff
@@ -0,0 +1,133 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type TargetType string
+
+const (
+	TargetTypeUpstreamTunnel TargetType = "UpstreamTunnel"
+	TargetTypeUI             TargetType = "UI"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Namespaced
+
+type TLSTerminatedRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              TLSTerminatedRouteSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// TLSTerminatedRouteList contains a list of ManagedCluster resources.
+type TLSTerminatedRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []TLSTerminatedRoute `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type TLSPassThroughRoute struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Dest is the destination URL
+	Spec TLSPassThroughRouteSpec `json:"spec"`
+}
+
+type TLSPassThroughRouteSpec struct {
+	// +kubebuilder:validation:Enum=UpstreamTunnel
+	// +required
+	Target TargetType `json:"target"`
+
+	// SNIMatch is used to match requests based on the server name for the intended destination server. Matching requests
+	// will be proxied to the Destination.
+	// +required
+	SNIMatch *SNIMatch `json:"sniMatch"`
+
+	// Destination is the destination url to proxy the request to.
+	// +required
+	Destination string `json:"destination"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type TLSPassThroughRouteList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []TLSPassThroughRoute `json:"items"`
+}
+
+type TLSTerminatedRouteSpec struct {
+	// +kubebuilder:validation:Enum=UpstreamTunnel;UI
+	Target TargetType `json:"target"`
+
+	// PathMatch is used to match requests based on what's in the path. Matching requests will be proxied to the Destination
+	// defined in this structure.
+	// +required
+	PathMatch *PathMatch `json:"pathMatch"`
+
+	// Destination is the destination URL where matching traffic is routed to.
+	// +required
+	Destination string `json:"destination"`
+
+	// CABundle is where we read the CA bundle from to authenticate the
+	// destination (if non-empty)
+	// +required
+	CABundle *v1.ConfigMapKeySelector `json:"caBundle,omitempty"`
+
+	// MTLS settings, and if both are specified MTLs is used.
+	// +optional
+	MTLSCert *v1.SecretKeySelector `json:"mtlsCert,omitempty"`
+
+	// +optional
+	MTLSKey *v1.SecretKeySelector `json:"mtlsKey,omitempty"`
```

Comment:
I think for MTLSCert and MTLSKey we probably want to phrase these in terms of where they are used. Do these represent the ClientCert and ClientKey used by Voltron on the forwarded connection?

We should at least call that out in the comment, although I'd probably go so far as to rename them as well to make it more clear that these are client secrets on the forwarded connection (and not used at all on the inbound connection).

Reply:
Updated the names and descriptions